crates/ruma-client-api/src/r0/state/get_state_events_for_key.rs

```diff
@@ -112,35 +112,28 @@ impl ruma_api::IncomingRequest for IncomingRequest {
 
     const METADATA: ruma_api::Metadata = METADATA;
 
-    fn try_from_http_request<T: AsRef<[u8]>>(
-        request: http::Request<T>,
+    fn try_from_http_request<T: AsRef<[u8]>, S: AsRef<str>>(
+        _request: http::Request<T>,
+        path_args: &[S],
     ) -> Result<Self, ruma_api::error::FromHttpRequestError> {
-        use std::convert::TryFrom;
-
-        let path_segments: Vec<&str> = request.uri().path()[1..].split('/').collect();
-
-        let room_id = {
-            let decoded =
-                percent_encoding::percent_decode(path_segments[4].as_bytes()).decode_utf8()?;
-
-            Box::<RoomId>::try_from(&*decoded)?
-        };
-
-        let event_type = {
-            let decoded =
-                percent_encoding::percent_decode(path_segments[6].as_bytes()).decode_utf8()?;
-
-            EventType::try_from(&*decoded)?
-        };
-
-        let state_key = match path_segments.get(7) {
-            Some(segment) => {
-                let decoded = percent_encoding::percent_decode(segment.as_bytes()).decode_utf8()?;
-
-                String::try_from(&*decoded)?
-            }
-            None => "".into(),
-        };
+        let (room_id, event_type, state_key): (Box<RoomId>, EventType, String) =
```

Comment:
I wonder, would this work without the branch if `state_key` was an `Option<String>` (which you could then `.unwrap_or("")`)?

Reply:
I'm not entirely sure how smart serde is with that, so I avoided it

Reply:
It doesnt seem to be possible;

<details>
<summary>Code</summary>

```rs
use serde::{Serialize,Deserialize};
use serde::de::value::{SeqDeserializer,Error};

#[derive(Serialize, Debug)]
struct Example(String,String,Option<String>);

fn main() {
    let path_args: &[String] = &["".into(), "".into()];
    
    thing(path_args);
}

fn thing<S>(path_args: &[S])
where
    S: AsRef<str>,
{
    let (a, b, c,) = Deserialize::deserialize(SeqDeserializer::<_, Error>::new(path_args.iter().map(AsRef::as_ref))).unwrap();
    
    let s = Example(a, b, c);
    
    dbg!(s);
}
```

</details>

`thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error("invalid length 2, expected a tuple of size 3")'`

Reply:
From chat; i'll leave a FIXME in both cases here